lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -0,0 +1,1288 @@
+  if (!singleIndirectionLevel(elementType)) {
+    assert(false && "FIXME: invalid element type");
+    return {};
+  }
+  return Base::get(elementType.getContext(), FIR_POINTER, elementType);
+}
+
+mlir::Type fir::PointerType::getEleTy() const {
+  return getImpl()->getElementType();
+}
+
+mlir::LogicalResult fir::PointerType::verifyConstructionInvariants(
+    const mlir::AttributeStorage *context, mlir::Type eleTy) {
+  if (eleTy.isa<BoxType>() || eleTy.isa<BoxCharType>() ||
+      eleTy.isa<BoxProcType>() || eleTy.isa<DimsType>() ||
+      eleTy.isa<FieldType>() || eleTy.isa<LenType>() || eleTy.isa<HeapType>() ||
+      eleTy.isa<PointerType>() || eleTy.isa<ReferenceType>() ||
+      eleTy.isa<TypeDescType>())
+    return mlir::failure();
+  return mlir::success();
+}
+
+// Heap<T>
+
+HeapType fir::HeapType::get(mlir::Type elementType) {
+  if (!singleIndirectionLevel(elementType)) {
+    assert(false && "FIXME: invalid element type");
+    return {};
+  }
+  return Base::get(elementType.getContext(), FIR_HEAP, elementType);
+}
+
+mlir::Type fir::HeapType::getEleTy() const {
+  return getImpl()->getElementType();
+}
+
+mlir::LogicalResult
+fir::HeapType::verifyConstructionInvariants(const mlir::AttributeStorage *,
+                                            mlir::Type eleTy) {
+  if (eleTy.isa<BoxType>() || eleTy.isa<BoxCharType>() ||
+      eleTy.isa<BoxProcType>() || eleTy.isa<DimsType>() ||
+      eleTy.isa<FieldType>() || eleTy.isa<LenType>() || eleTy.isa<HeapType>() ||
+      eleTy.isa<PointerType>() || eleTy.isa<ReferenceType>() ||
+      eleTy.isa<TypeDescType>())
+    return mlir::failure();
+  return mlir::success();
+}
+
+// Sequence<T>
+
+SequenceType fir::SequenceType::get(const Shape &shape, mlir::Type elementType,
+                                    mlir::AffineMapAttr map) {
+  auto *ctxt = elementType.getContext();
+  return Base::get(ctxt, FIR_SEQUENCE, shape, elementType, map);
+}
+
+mlir::Type fir::SequenceType::getEleTy() const {
+  return getImpl()->getElementType();
+}
+
+mlir::AffineMapAttr fir::SequenceType::getLayoutMap() const {
+  return getImpl()->getLayoutMap();
+}
+
+SequenceType::Shape fir::SequenceType::getShape() const {
+  return getImpl()->getShape();
+}
+
+mlir::LogicalResult fir::SequenceType::verifyConstructionInvariants(
+    const mlir::AttributeStorage *attributeStorage,
+    const SequenceType::Shape &shape, mlir::Type eleTy,
+    mlir::AffineMapAttr map) {
+  // DIMENSION attribute can only be applied to an intrinsic or record type
+  if (eleTy.isa<BoxType>() || eleTy.isa<BoxCharType>() ||
+      eleTy.isa<BoxProcType>() || eleTy.isa<DimsType>() ||
+      eleTy.isa<FieldType>() || eleTy.isa<LenType>() || eleTy.isa<HeapType>() ||
+      eleTy.isa<PointerType>() || eleTy.isa<ReferenceType>() ||
+      eleTy.isa<TypeDescType>() || eleTy.isa<SequenceType>()) {
+    auto context = attributeStorage->getDialect().getContext();
+    mlir::emitError(mlir::UnknownLoc::get(context),
```

Comment:
emitError returns a diagnostic, which converts to `mlir::failure` automatically, so you can just return that directly.